### PR TITLE
meta: support for metadata links (CRAFT-370)

### PR DESCRIPTION
**Pull request by sergiusens**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/meta/snap.py [hunk 1]
```diff
@@ -317,6 +319,31 @@ def from_dict(cls, snap_dict: Dict[str, Any]) -> "Snap":
...
         layout = snap_dict.pop("layout", None)
         license = snap_dict.pop("license", None)
+
+        # snap.yaml will have links, snapcraft.yaml the top level entries.
+        # some considerations:
+        # - links should not be allowed in snapcraft.yaml by the schema.
+        # - the top level entries would never exist in snap.yaml
+        #   (snap pack would reject them).
+        # snap.yaml can have a contact entry at the top level which is equivalent
+        # to .links.contact[0] so even if snapcraft does not perfora a snap.yaml
```
Comment (cjp256) [inline snapcraft/internal/meta/snap.py:329]:
```suggestion
        # to .links.contact[0] so even if snapcraft does not perform a snap.yaml
```

File: snapcraft/internal/meta/snap.py [hunk 2]
```diff
@@ -317,6 +319,31 @@ def from_dict(cls, snap_dict: Dict[str, Any]) -> "Snap":
...
+        # to snap.yaml transformation today it would lead to a compatible result.
```
Comment (cjp256) [inline snapcraft/internal/meta/snap.py:330]:
```suggestion
        # to snap.yaml transformation today, it would lead to a compatible result.
```

**Review by cjp256 — APPROVED** [review]

File: snapcraft/internal/meta/snap.py [hunk 3]
```diff
@@ -317,6 +319,31 @@ def from_dict(cls, snap_dict: Dict[str, Any]) -> "Snap":
 
         layout = snap_dict.pop("layout", None)
         license = snap_dict.pop("license", None)
+
+        # snap.yaml will have links, snapcraft.yaml the top level entries.
+        # some considerations:
+        # - links should not be allowed in snapcraft.yaml by the schema.
+        # - the top level entries would never exist in snap.yaml
+        #   (snap pack would reject them).
```
Comment (pedronis) [inline snapcraft/internal/meta/snap.py:327]: we have no code to do that in my current PR, we tend not to refuse extra fields but maybe in this case we should

File: snapcraft/internal/meta/snap.py [hunk 4]
```diff
@@ -317,6 +319,31 @@ def from_dict(cls, snap_dict: Dict[str, Any]) -> "Snap":
...
+        # snap.yaml can have a contact entry at the top level which is equivalent
```
Comment (pedronis) [inline snapcraft/internal/meta/snap.py:328]: we don't support that ATM, current snapd would just ignore it, as I said new code would also ignore it

**Review by pedronis — COMMENTED** [review]
the code looks fine but the comment doesn't quite completely match current/pending reality on the snap.yaml side

File: snapcraft/internal/meta/snap.py (hunk 3 above)
Comment (sergiusens) [reply snapcraft/internal/meta/snap.py:327]: thanks, this is from the context of a snap.yaml generated by snapcraft. I will remove the "snap pack" will reject them comment as you are right, that slipped my mind

File: snapcraft/internal/meta/snap.py (hunk 4 above)
Comment (sergiusens) [reply snapcraft/internal/meta/snap.py:328]: this is from the context of snapcraft.yaml, where we never supported the contact field anyways and just a comment about an internal problem we have with transformations by handling snap.yaml and snapcraft.yaml with the same class.

**Comment by sergiusens** [comment]
Thanks @pedronis, I clarified the code comments

**Pull request merged**